# CSCN8010-Lab5
Lifei Wang 8855277

1. Utilize the diabetes dataset from lab 4. Perform cross-validation on nine polynomial models, ranging from degree 0 to 8. (2 points)

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import pyplot as plt
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline
from sklearn import preprocessing
from sklearn import datasets
from sklearn.metrics import r2_score, mean_absolute_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split

X, y = datasets.load_diabetes(as_frame=True, scaled=False, return_X_y=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
# construct 9 poly models from degree 0 to 8 using a function
# reference from chatGPT and cross-validation.ipynb from class notebook

# lists to store model metrics
degrees = []
r2_scores = []
mae_scores = []

for degree in range(9):
    # create poly features
    poly_features = PolynomialFeatures(degree=degree)
    X_poly = poly_features.fit_transform(X)

    # build linear regression model
    poly_reg_model = LinearRegression()

    # combine into pipeline
    model = make_pipeline(poly_features, poly_reg_model)

    # cross-validation
    # neg_mean_squared_error returns the negative of the mean squared error
    # cross_val_score is designed so that higher scores are better
    # a score of 0 is the best possible score, and values less than 0 indicate errors.
    scores = cross_val_score(model, X, y, cv=5, scoring='neg_mean_squared_error')

    # get MSE score
    mean_score = -scores.mean()

    # get R-squared score
    y_pred = cross_val_predict(model, X, y, cv=5)
    r2 = r2_score(y, y_pred)

    # get MAE
    mae = mean_absolute_error(y, y_pred)

    # store in lists
    degrees.append(degree)
    r2_scores.append(r2)
    mae_scores.append(mae)

2. Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. 

    Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. 

    Calculate the mean value and standard deviation of these metrics from the cross-validation. 
    
    Include both values. (2 points)

In [13]:
from tabulate import tabulate

# create a table
table = []

table.append(['Degree', 'R-Squared', 'Mean Absolute Error'])

for degree, r2, mae in zip(degrees, r2_scores, mae_scores):
    table.append([degree, r2, mae])

# get mean value and sd of metrics
r2_mean = np.mean(r2_scores)
r2_sd = np.std(r2_scores)
mae_mean = np.mean(mae_scores)
mae_sd = np.std(mae_scores)

# append mean and sd values to table
table.append(['Mean', r2_mean, mae_mean])
table.append(['Standard Deviation', r2_sd, mae_sd])

print(tabulate(table, headers='firstrow', tablefmt='fancy_grid'))

╒════════════════════╤═════════════════╤═══════════════════════╕
│ Degree             │       R-Squared │   Mean Absolute Error │
╞════════════════════╪═════════════════╪═══════════════════════╡
│ 0                  │     -0.00882372 │               66.0393 │
├────────────────────┼─────────────────┼───────────────────────┤
│ 1                  │      0.495322   │               44.2749 │
├────────────────────┼─────────────────┼───────────────────────┤
│ 2                  │     -0.58401    │               66.8434 │
├────────────────────┼─────────────────┼───────────────────────┤
│ 3                  │   -195.308      │              341.371  │
├────────────────────┼─────────────────┼───────────────────────┤
│ 4                  │   -572.362      │              656.629  │
├────────────────────┼─────────────────┼───────────────────────┤
│ 5                  │   -446.482      │              563.02   │
├────────────────────┼─────────────────┼───────────────────────┤
│ 6                  │  -

4. Optional (No Extra Grade):
    
    Additional analysis and interpretation of the models' performances are welcome. 
    
    You may explore further insights beyond the required metrics. Feel free to include your findings.

3. Identification of the Best Model: 

    Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. 
    
    Provide an explanation for choosing this specific model. (2 points)

From the table above, the best model should have the highest R-squared value and lowest MAE value.


Therefore, the best model would be the model of degree=1 with R-squared value=0.495322 and MAE value=44.2749

Since R-squared measures the proportion of variance in the model, a higher R-squared value indicates a larger proportion of variability that the model could explain which means better performance.

MAE measures the average magnitude of errors of the model, a lower MAE indicates more accuracy of the model and better performance.